lib/ts/BufferWriter.h

```diff
@@ -0,0 +1,309 @@
+#if !defined TS_BUFFERWRITER_H_
+#define TS_BUFFERWRITER_H_
+
+/** @file
+
+    Utilities for generating character sequences.
+
+    @section license License
+
+    Licensed to the Apache Software Foundation (ASF) under one
+    or more contributor license agreements.  See the NOTICE file
+    distributed with this work for additional information
+    regarding copyright ownership.  The ASF licenses this file
+    to you under the Apache License, Version 2.0 (the
+    "License"); you may not use this file except in compliance
+    with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+ */
+
+#include <utility>
+#include <cstring>
+
+#include <ts/string_view.h>
+#include <ts/ink_assert.h>
+
+namespace ts
+{
+// Abstract class.
+//
+class BufferWriter
+{
+protected:
+  // The _pushBack() functions "add" characters at the end.  If these functions discard any characters, this must put the instance
+  // in an error state (indicated by the override of error() ).  Derived classes must not assume the _pushBack() functions will
+  // not be called when the instance is in an error state.
+
+  virtual void _pushBack(char c) = 0;
+
+  virtual void
+  _pushBack(string_view sV)
+  {
+    for (size_t i = 0; i < sV.size(); ++i) {
+      _pushBack(sV[i]);
+    }
+  }
+
+public:
+  // Returns true if the instance is in an error state.
+  //
+  virtual bool error() const = 0;
+
+  BufferWriter &
+  c(const char c)
```

Comment:
The terse abbreviation US is good if it means United States, and bad if it means Upper Silesia.  Terse abbreviations are advantageous if they are used very frequently.   After doing some work in proxy/http, I had difficulty working with all the raw char pointers and old <cstring> functions.  I'm looking for some kind of utilities to wrap that low-level stuff in.  If we  had some good wrapper utilities, code like this would (hopefully) gradually convert over to using them.  Since they would be used a lot, short names would be advantageous.